File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode *l1=list1,*l2=list2;
        ListNode* res = new ListNode(0);
        ListNode* curr = res;
        while(l1 != NULL && l2 != NULL){
            if (l1->val <= l2->val) {
                curr->next = l1; // Reuse the node from l1
                l1 = l1->next;  // Move to the next node in l1
            } else {
                curr->next = l2; // Reuse the node from l2
                l2 = l2->next;  // Move to the next node in l2
            }
            curr = curr->next; // Advance the result list
        }

        if(l1 !=NULL){
            curr->next = l1;
        }else{
            curr->next = l2;
        }

        ListNode* mergedll = res->next;
        delete res;
        return mergedll;
    }
};

File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        if not list1:
            return list2
        if not list2:
            return list1
        
        h1 = list1
        h2 = list2
        dummy = ListNode(0)
        curr = dummy
        while h1 and h2:
            if h1.val <= h2.val:
                curr.next = h1
                h1 = h1.next
            else:
                curr.next = h2
                h2 = h2.next
            curr = curr.next
        
        if h1:
            curr.next = h1
        else:
            curr.next = h2
            
        return dummy.next